[ML] Add log structure finder functionality

droberts195 opened the pull request:
This change adds a library to ML that can be used to deduce a log
file's structure given only a sample of the log file.

Eventually this will be used to add an endpoint to ML to make the
functionality available to end users, but this will follow in a
separate change.

The functionality is split into a library so that it can also be
used by a command line tool without requiring the command line
tool to include all server code.

droberts195 commented:
I've marked this `>non-issue` as it's invisible to the end user in this state.  It will get added to the release notes when the corresponding endpoint is added.

benwtrent commented on x-pack/plugin/ml/log-structure-finder/src/main/java/org/elasticsearch/xpack/ml/logstructure/TimestampFormatFinder.java:
```diff
@@ -0,0 +1,422 @@
+
+        @Override
+        public boolean equals(Object other) {
+            if (other == null) {
```
#nit add the equal object predicate.  
```
if (this == other) {
    return true;
}
```

droberts195 commented on x-pack/plugin/ml/log-structure-finder/src/main/java/org/elasticsearch/xpack/ml/logstructure/GrokPatternCreator.java:
```diff
@@ -0,0 +1,591 @@
+        boolean wildcardRequired = true;
+        for (int i = 0; i < driver.length(); ++i) {
+            char ch = driver.charAt(i);
+            if (PUNCTUATION_OR_SPACE.indexOf(ch) >= 0 && others.stream().allMatch(other -> other.indexOf(ch) >= 0)) {
```
I removed `PUNCTUATION_OR_SPACE` and `NEEDS_ESCAPING` and replaced them with a `HashMap` that enables both things to be checked with one lookup.

dimitris-athanasiou commented on x-pack/plugin/ml/log-structure-finder/src/main/java/org/elasticsearch/xpack/ml/logstructure/GrokPatternCreator.java:
```diff
@@ -0,0 +1,600 @@
+        }
+
+        /**
+         * This must only be called if {@link #matchesAll} returns <code>true</code>.
```
As this is only called in one place and right after `matchesAll` was called, I think we could move the `matchesAll` check in this method and so remove the risk of letting it be called without fulfilling the `matchesAll` contract.

dimitris-athanasiou commented on x-pack/plugin/ml/log-structure-finder/src/main/java/org/elasticsearch/xpack/ml/logstructure/GrokPatternCreator.java:
```diff
@@ -0,0 +1,600 @@
+     * matched section and the pieces before and after it.  Recurse to find more matches in the pieces
+     * before and after and update the supplied string builder.
+     */
+    private static void processCandidateAndSplit(List<String> explanation, Map<String, Integer> fieldNameCountStore,
```
The long arg lists in these methods make me wonder if we could refactor as an object whose members capture those parameters. Hard to come with a more concrete suggestion without trying to do it, but these long arg lists and static methods set make me wonder if we could simplify by going more OO here. This class seems like it is doing a lot and could be split up in several smaller classes with reduced responsibilities.

droberts195 replied:
Good idea - it's done in https://github.com/elastic/elasticsearch/pull/32788/commits/37a92f7a67f762fbfbbc92d53db807761b5b0ef2

This class got progressively more complex over time.  When I started out it made sense to just have a couple of static methods but now you're right it needs to be an object with members.

benwtrent commented on x-pack/plugin/ml/log-structure-finder/src/main/java/org/elasticsearch/xpack/ml/logstructure/SeparatedValuesLogFileStructureFinder.java:
```diff
@@ -0,0 +1,458 @@
+        if (rows.size() < 3) {
+            explanation.add("Too little data to accurately assess whether header is in sample - guessing it is");
+        } else {
+            isCsvHeaderInFile = isFirstRowUnusual(explanation, rows);
```
So, we want to automatically determine if a CSV header is given in the sample or not. Here are a couple of quick checks that we may want to try before doing expensive statistical checking and Levenshtein distances. Of course, this depends on how flexible we want to be. 

- Are all entries unique in the header row? You should not have duplicate header entries
- Do we want to support `null` headers fields? If not, it stands to reason that none should be null in the sample
- Are all entries actual `string` values? I am not sure it is valid to have a header that is just a number or a blank string.

droberts195 replied:
The duplicates test is added in https://github.com/elastic/elasticsearch/pull/32788/commits/cf2d6750c29069b358dac7e30702e99a52a73eae

droberts195 commented on x-pack/plugin/ml/log-structure-finder/src/main/java/org/elasticsearch/xpack/ml/logstructure/AbstractLogFileStructureFinder.java:
```diff
@@ -0,0 +1,81 @@
+    private static final int KEYWORD_MAX_LEN = 256;
+    private static final int KEYWORD_MAX_SPACES = 5;
+
+    protected static Map<String, String> guessScalarMapping(List<String> explanation, String fieldName, Collection<String> fieldValues) {
```
It would be nicer in some ways, but it would cause quite a lot of side effects.  At the moment `explanation` is built up by 3 pieces of code:

1. The character set finder
2. The high level structure determination (i.e. the `*LogFileStructure::canCreateFromSample()` methods)
3. The detailed structure determination

The unit tests test static methods that update explanations without having to create concrete objects.  And the unit tests also print the explanation or partial explanation for whatever they've tested, which means that for unit tests we want the explanation to be owned by the test code.

I think one solution would be to introduce an `Explanation` class that encapsulates the list of strings.  Then that could be a member of `AbstractLogFileStructureFinder`, which would avoid passing it around the member functions of that hierarchy so much.  But it would be documented that other classes might be adding to the same explanation.

droberts195 replied:
It's also tricky because the `GrokPatternCreator` needs to find mappings, relying on the fact that `guessScalarMapping()` is `static`.

droberts195 commented on x-pack/plugin/ml/log-structure-finder/src/main/java/org/elasticsearch/xpack/ml/logstructure/AbstractLogFileStructureFinder.java:
```diff
@@ -0,0 +1,81 @@
+            }
+        }
+
+        else if (fieldValues.stream().allMatch(IP_GROK::match)) {
```
It should be `else if`, but I'll remove the accidental extra blank line.

droberts195 commented on x-pack/plugin/ml/log-structure-finder/src/main/java/org/elasticsearch/xpack/ml/logstructure/AbstractLogFileStructureFinder.java:
```diff
@@ -0,0 +1,81 @@
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.ml.logstructure;
```
OK cool.  I renamed the package to `logstructurefinder`.

dimitris-athanasiou commented on x-pack/plugin/ml/log-structure-finder/src/main/java/org/elasticsearch/xpack/ml/logstructure/LogFileStructure.java:
```diff
@@ -0,0 +1,614 @@
+        public LogFileStructure build() {
+
+            if (numLinesAnalyzed <= 0) {
+                throw new IllegalArgumentException("Number of lines analyzed must be positive.");
```
nit: In other classes we have such validations in the corresponding builder setter. Arguably there are different benefits doing it either way. Just raising to see what we think about the 2 ways.

droberts195 replied:
I put it here because the most likely error is failing to call the builder setter at all.  This error would occur during development/adding a new log file structure.  It won't be encountered by end users because they should never be building one of these objects.

dimitris-athanasiou commented on x-pack/plugin/ml/log-structure-finder/src/main/java/org/elasticsearch/xpack/ml/logstructure/LogFileStructure.java:
```diff
@@ -0,0 +1,614 @@
+        public LogFileStructure build() {
+
+            if (numLinesAnalyzed <= 0) {
+                throw new IllegalArgumentException("Number of lines analyzed must be positive.");
```
Another inconsistency here is whether the error messages contain a natural description of the field vs. the actual field name.

droberts195 replied:
I think it's academic in this case, because it's our code supplying the values rather than the end user.  So this check won't trigger as a result of a user calling a REST endpoint unless there's a bug in our code.

dimitris-athanasiou commented on x-pack/plugin/ml/log-structure-finder/src/main/java/org/elasticsearch/xpack/ml/logstructure/JsonLogFileStructureFinder.java:
```diff
@@ -0,0 +1,76 @@
+
+    JsonLogFileStructureFinder(List<String> explanation, String sample, String charsetName, Boolean hasByteOrderMarker) throws IOException {
+
+        List<Map<String, ?>> sampleRecords = new ArrayList<>();
```
It sees that the Finder classes' constructors are doing the bulk of the work here. I wonder if it would be best to avoid that by introducing a method that clearly communicates that processing is going to occur.

droberts195 replied:
I thought about this, but then the member variables cannot be `final`, and if you have one of these objects there's a potential doubt about whether the processing method has been called or not.

dimitris-athanasiou replied:
Good point. The other idea would be to go with static factory methods. But I don't feel strongly about it.

droberts195 replied:
Good idea - I moved the major chunks of processing to static factory methods in https://github.com/elastic/elasticsearch/pull/32788/commits/baf863f94844b5b6ef5895437ab11cb6e997f3b1

droberts195 commented on x-pack/plugin/ml/log-structure-finder/src/main/java/org/elasticsearch/xpack/ml/logstructure/AbstractStructuredLogFileStructureFinder.java:
```diff
@@ -0,0 +1,140 @@
+        for (Tuple<String, TimestampMatch> firstSampleMatch : firstSampleMatches) {
+
+            boolean allGood = true;
+            for (Map<String, ?> sampleRecord : sampleRecords.subList(1, sampleRecords.size())) {
```
We set `allGood = true` before the loop and in the case of only one sample it's fine if the loop does nothing at all since we know the tests will pass against the first sample.  So I removed the single sample optimization to enable `findCandidates()` to be extracted, but left this loop starting with the second sample.

droberts195 commented on x-pack/plugin/ml/log-structure-finder/src/main/java/org/elasticsearch/xpack/ml/logstructure/LogFileStructure.java:
```diff
@@ -0,0 +1,614 @@
+    private final String grokPattern;
+    private final List<String> timestampFormats;
+    private final String timestampField;
+    private final boolean needClientTimezone;
```
Because it makes sense to include this in the output in all cases, whereas for some of the other boolean flags they only make sense when a particular high level structure has been found.  For example, `has_header_row` does not make sense at all for JSON.

droberts195 commented on x-pack/plugin/ml/log-structure-finder/src/main/java/org/elasticsearch/xpack/ml/logstructure/AbstractStructuredLogFileStructureFinder.java:
```diff
@@ -0,0 +1,140 @@
+            for (Map<String, ?> sampleRecord : sampleRecords) {
+                for (String fieldName : sampleRecord.keySet()) {
+                    mappings.computeIfAbsent(fieldName, key -> guessMapping(explanation, fieldName,
+                        sampleRecords.stream().flatMap(record -> {
```
I think that would negate the benefit of using `computeIfAbsent()`.  This is in an inner loop and if 1000 records all contain the same fields we don't want to do the complex processing 1000 times more than necessary.

If it's really hard to read as it stands I guess an alternative approach would be to first build a `HashSet` of all the unique field names across all sample records, then iterate through that to drive creation of the mappings for each field.

droberts195 commented on x-pack/plugin/ml/log-structure-finder/src/main/java/org/elasticsearch/xpack/ml/logstructure/AbstractStructuredLogFileStructureFinder.java:
```diff
@@ -0,0 +1,140 @@
+
+public abstract class AbstractStructuredLogFileStructureFinder extends AbstractLogFileStructureFinder {
+
+    protected Tuple<String, TimestampMatch> guessTimestampField(List<String> explanation, List<Map<String, ?>> sampleRecords) {
```
While adding the comment I noticed that for ordered structured formats with multiple timestamps we weren't consistently taking the first timestamp in left-to-right order in the original data (due to using a `HashMap` rather than a `LinkedHashMap`).  So I fixed that too.

davidkyle commented on x-pack/plugin/ml/log-structure-finder/src/main/java/org/elasticsearch/xpack/ml/logstructure/LogFileStructureFinderManager.java:
```diff
@@ -0,0 +1,232 @@
+                for (int i = 0; i < spaceBytes.length && spaceEncodingContainsZeroByte == false; ++i) {
+                    spaceEncodingContainsZeroByte = (spaceBytes[i] == 0);
+                }
+                if (containsZeroBytes && spaceEncodingContainsZeroByte == false) {
```
I'm learning a lot about character encoding

davidkyle commented on x-pack/plugin/ml/log-structure-finder/src/main/java/org/elasticsearch/xpack/ml/logstructure/SeparatedValuesLogFileStructureFinder.java:
```diff
@@ -0,0 +1,458 @@
+
+        Tuple<Boolean, String[]> headerInfo = findHeaderFromSample(explanation, rows);
+        boolean isCsvHeaderInFile = headerInfo.v1();
+        String[] csvHeader = headerInfo.v2();
```
nit `delineatedHeader` or `separatedValuesHeader`

droberts195 replied:
I changed it to just `header`.  `csv` still appears in a few places because SuperCSV has the same problem of using "CSV" in its name while being a generic delimited values parser.

droberts195 commented on x-pack/plugin/ml/log-structure-finder/src/main/java/org/elasticsearch/xpack/ml/logstructure/LogFileStructure.java:
```diff
@@ -0,0 +1,614 @@
+                    if (shouldTrimFields != null) {
+                        throw new IllegalArgumentException("Should trim fields may not be specified for [" + format + "] structures.");
+                    }
+                    // $FALL-THROUGH$
```
It does in Eclipse - see 
https://stackoverflow.com/a/13145718

However, I had to add the `@SuppressWarnings` on the whole method as the Gradle build runs with `-WError`, so this isn’t actually necessary to suppress the warning. I can change it to a more human-readable `// Fallthrough` if you prefer.

droberts195 commented:
I think I've now addressed all the review comments so far.

The only thing I'm left wondering is whether after https://github.com/elastic/elasticsearch/pull/32788/commits/baf863f94844b5b6ef5895437ab11cb6e997f3b1 it would make sense to get rid of the `abstract` base classes and move the methods (which are all now `static`) into a new `LogStructureUtils` class, a it's a bit weird to have `abstract` base classes with only `static` methods.  I could also move a `static` method from the `GrokPatternCreator` into the new utility class to break a circular dependency with the `TextLogStructureFinder`.

I think I'll make this change tomorrow unless anyone has a better idea and then hopefully the PR will be in a fit state to merge.

droberts195 commented:
@elasticmachine test this please
